### PR TITLE
GH-43046: [C++] Fix avx2 gather rows more than 2^31 issue in `CompareColumnsToRows`

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

See #43046.

### What changes are included in this PR?

Use unsigned offset safe gather introduced in #42188 which is to fix similar issues.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

None.

* GitHub Issue: #43046

### CODE REVIEW

File: cpp/src/arrow/compute/row/compare_test.cc [hunk 1]
```diff
@@ -300,6 +300,139 @@ TEST(KeyCompare, CompareColumnsToRowsLarge) {
               num_rows);
   }
 }
+
+// Compare columns to rows of row ids larger than 2^31 within a row table.
+// Certain AVX2 instructions may behave unexpectedly causing troubles like GH-43046.
+TEST(KeyCompare, CompareColumnsToRowsMany) {
```
Comment (zanmato1984) [inline cpp/src/arrow/compute/row/compare_test.cc:306]: This case takes about 3min to run - it has to construct a row table with more than 2^31 rows which is necessary to reproduce the issue. Is it proper to add it here? Or somewhere else that is more suitable for long-running tests? @pitrou please help?

**Comment by zanmato1984** [comment]
Some CI failure seems related. Looking. And turn to draft before figure that out.

**Comment by zanmato1984** [comment]
> Some CI failure seems related. Looking. And turn to draft before figure that out.

No big deal, changed one more line by mistake:
[`c1eb593` (#43065)](https://github.com/apache/arrow/pull/43065/commits/c1eb593df0fcdd395811c2d37768a77bbbb0c38d#diff-4dcfb2dd9add2d770dcbd924498120ab30e38186dd4bdc60708485c31074ce67L243)

Fixed. Re-ready for review.

File: cpp/src/arrow/compute/row/compare_test.cc (hunk 1 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc:306]: > Or somewhere else that is more suitable for long-running tests?

For example some compiler flags that will be only turned on for time consuming tests? Like `ARROW_BUILD_DETAILED_BENCHMARKS` for benchmarks.

Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_test.cc:306]: We don't have such a flag, no. Hopefully we can shrink the test execution time?

Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_test.cc:306]: Also, how much RAM does this require?

Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc:306]: > Also, how much RAM does this require?

It takes 24+GB virtual memory.

Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_test.cc:306]: 24+GB memory is unpractical for most CI configurations (which are typically based on small VMs), and also many local dev environments. So, I think we'll unfortunately have to disable this test, perhaps using the `LARGE_MEMORY_TEST` annotation (you can grep through the sources to look for it).

Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc:306]: OK, that's unfortunate but seems to be the only option. Shall I now stop my (hopeless) attempts to speedup the test then?

Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_test.cc:306]: If speeding up the test requires consuming even more memory, then it doesn't sound like a good idea indeed.

**Comment by zanmato1984** [comment]
My investigation on how much RAM this test uses disclosed something new. The test was a fake pass. There are offset overflows (of `uint32_t`) happening - I'll open up an issue for it later. But since the number of rows and the size of the `row_table_right` is multiple of `columns_left`, the overflowed offsets slide to the beginning of the row table, making the comparison actually happens between the first `num_rows_batch` rows in row table and the batch. Simply reducing the number of rows of the batch by 1 will make the offset overflow bad enough to fail the test:
```diff
-  constexpr int64_t num_rows_batch = std::numeric_limits<uint16_t>::max() + 1ll;
+  constexpr int64_t num_rows_batch = std::numeric_limits<uint16_t>::max();
```

The further impact to this PR is that, it is technically impossible to construct a valid test case to reproduce the issue that this PR is aiming to solve. This issue requires the following conditions to reproduce:
1. A fixed length column exists;
2. A var length column exists;
3. The number of rows is greater than 2^31.

No matter how small the fixed length column and var length column's content are, there is an 32b "offset" (of the var length column) within the row. And more than 2^31 rows effectively take 2^31 * 4B > 4GB data, then the offset within the row must overflow. Or in other words, one could not simply construct a row table of more than 2^31 rows WITHOUT overflowing the row offset.

We can say that the issue that this PR is for won't actually happen, because something even worse (the overflow) must have happened way earlier. But I still want to have this fix in because the fix is simple, safe, without overhead, and can prevent bad things from happening in the future. (maybe when someday we improve the row table to allow valid bigger-than-2^31 rows?)

What's your opinion? @pitrou  Thanks.

**Comment by pitrou** [comment]
Two things:
1) Yes, I agree we should fix the bug even though we can't write a unit test to avoid regressions
2) Are you saying the current fix is not enough, or am I misunderstanding you?

**Comment by zanmato1984** [comment]
> 2. Are you saying the current fix is not enough, or am I misunderstanding you?

No, the current fix is perfectly enough to solely solve the "gather data from offsets for rows[>=2^31]" issue as long as it actually happens.

What I was talking about is an "overflow" issue will always happen way before the gather issue could actually happen. And once we fixed the overflow issue (I'm going to introduce explicit error), the gather issue won't happen too because the run will be interrupted by my presumable explicit error. But, quoting myself:
> and can prevent bad things from happening in the future (maybe when someday we improve the row table to allow valid bigger-than-2^31 rows?).

**Comment by zanmato1984** [comment]
The test has been removed from this PR. But it will come back, with a slightly modification and different coverage intention, when this TODO is being mitigated:
https://github.com/apache/arrow/pull/43065/files#diff-26ce8f4fb451b09705a8cadde87fbcc588d0cfe05a91b5677fbda95af215218cR328

File: cpp/src/arrow/compute/row/row_internal.cc [hunk 2]
```diff
@@ -325,6 +325,7 @@ Status RowTableImpl::AppendSelectionFrom(const RowTableImpl& from,
     // Varying-length rows
     auto from_offsets = reinterpret_cast<const uint32_t*>(from.offsets_->data());
     auto to_offsets = reinterpret_cast<uint32_t*>(offsets_->mutable_data());
+    // TODO: The following two variables are possibly overflowing.
```
Comment (zanmato1984) [inline cpp/src/arrow/compute/row/row_internal.cc:328]: This is where the "overflow" happens in my other comments. I'll file another issue/PR for it.

**Comment by zanmato1984** [comment]
@pitrou Shall we move on with this? Thanks.

File: cpp/src/arrow/compute/row/row_internal.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/compute/row/row_internal.cc:328]: Did you file the other issue? If so, can you update the TODO here to the proper GH issue number? Thanks!

**Review by pitrou — CHANGES_REQUESTED** [review]
LGTM except one change to make

**Comment by pitrou** [comment]
@zanmato1984 Sorry for the delay, I was away sick

File: cpp/src/arrow/compute/row/row_internal.cc (hunk 2 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/row_internal.cc:328]: No. I was planning to file it after and refer to this TODO in it.

**Comment by zanmato1984** [comment]
> @zanmato1984 Sorry for the delay, I was away sick

You have my best wishes for you to take care and get better my friend.

**Comment by zanmato1984** [comment]
Darn, every time! Sorry for the trouble. I might need some time to get my branch right...

**Comment by zanmato1984** [comment]
> @zanmato1984 Also, can you rebase?

Rebase done.

Also, if you think it's better to file the other issue ahead of this PR and have the TODO pointing to it, I'll do that. Thank you!

**Comment by pitrou** [comment]
> Also, if you think it's better to file the other issue ahead of this PR and have the TODO pointing to it, I'll do that. Thank you!

Yes, that would be nice. Thanks!

**Comment by zanmato1984** [comment]
> > Also, if you think it's better to file the other issue ahead of this PR and have the TODO pointing to it, I'll do that. Thank you!
> 
> Yes, that would be nice. Thanks!

#43202 filed and assigned to the TODO in the PR.

**Review by pitrou — APPROVED** [review]

**Pull request merged**